appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+ */
+Genetics.Cage.EVENT_PROPERTY_NAMES = {
+  'ADD': ['TYPE', 'MOUSE', 'PLAYER_NAME'],
+  'START_GAME': [],
+  'FIGHT': ['TYPE', 'ID', 'RESULT', 'OPT_OPPONENT'],
+  'MATE': ['TYPE', 'ID', 'RESULT', 'OPT_PARTNER', 'OPT_OFFSPRING'],
+  // A mouse dies after using all fight and mate opportunities.
+  'RETIRE': ['TYPE', 'ID'],
+  // The oldest mouse dies when there is no room for it.
+  'OVERPOPULATION': ['TYPE', 'ID'],
+  // A mouse dies if it throws an exception.
+  'EXPLODE': ['TYPE', 'ID', 'SOURCE', 'OPT_CAUSE'],
+  // A mouse dies if it does not complete execution.
+  'SPIN': ['TYPE', 'ID'],
+  'END_GAME': ['TYPE', 'CAUSE', 'PICK_FIGHT_WINNER', 'CHOOSE_MATE_WINNER',
+    'MATE_ANSWER_WINNER']
+};
+
+/**
+ * Creates an event.
+ * @param {string } type The type of event.
+ * @param {...string|number|Genetics.Mouse} var_args The properties on
+ * the event.
+ * @constructor
+ */
+Genetics.Cage.Event = function(type, var_args) {
+  var template = Genetics.Cage.EVENT_PROPERTY_NAMES[type];
+  for (var i = 0; i < template.length; i++) {
+    this[template[i]] = arguments[i];
+  }
+};
+
+/**
+ * Adds the event to the queue depending on the mode the game is running as.
+ */
+Genetics.Cage.Event.prototype.addToQueue = function() {
+  // Headless does not need to keep track of events that are not END_GAME.
+  if (Genetics.Cage.HEADLESS &&
+      this.TYPE != Genetics.Cage.EVENT_PROPERTY_NAMES.END_GAME) {
+    return;
+  }
+  Genetics.Cage.EVENTS.push(this);
+};
+
+/**
+ * Time to end the battle.
+ * @type {number}
+ * @private
+ */
+Genetics.Cage.endTime_ = 0;
+
+/**
+ * Time limit of game (in milliseconds).
+ * @type {number}
+ */
+Genetics.Cage.GAME_TIME_LIMIT_MSC = 5 * 60 * 1000;
+
+/**
+ * Time limit for mouse function execution in number of interpreter steps (100k
+ * ticks runs for about 3 minutes).
+ * @type {number}
+ */
+Genetics.Cage.FUNCTION_TIMEOUT_STEPS = 100000;
+
+/**
+ * The next available player ID.
+ * @type {number}
+ * @private
+ */
+Genetics.Cage.nextAvailablePlayerId_ = 0;
+
+/**
+ * The next available mouse ID.
+ * @type {number}
+ * @private
+ */
+Genetics.Cage.nextAvailableMouseId_ = 0;
+
+/**
+ * Callback function for end of game.
+ * @type {Function}
+ * @private
+ */
+Genetics.Cage.doneCallback_ = null;
+
+
+/**
+ * Stop and reset the cage.
+ */
+Genetics.Cage.reset = function() {
+  clearTimeout(Genetics.Cage.pid);
+  Genetics.Cage.EVENTS.length = 0;
+  Genetics.Cage.aliveMice_.length = 0;
+  Genetics.Cage.miceMap_ = {};
+  // Cancel all queued life simulations.
+  for (var mouseId in Genetics.Cage.lifePidsMap_) {
+    if (Genetics.Cage.lifePidsMap_.hasOwnProperty(mouseId)) {
+      clearTimeout(Genetics.Cage.lifePidsMap_[mouseId]);
+    }
+  }
+  Genetics.Cage.lifePidsMap_ = {};
+  Genetics.Cage.nextAvailableMouseId_ = 0;
+};
+
+/**
+ * Adds a player to the game.
+ * @param {string} playerName The name of the player to be added.
+ * @param {string|!Function} code Player's code, or generator.
```

Comment:
Code is either a string or a function. Code retrieved from the blockly editor is a function.